icicle-data/src/Icicle/Common/Exp/Simp/ANormal.hs

```diff
@@ -59,65 +59,26 @@ anormal a_fresh xx
 anormalAllVars :: (Hashable n, Eq n) => a -> Exp (Ann a n) n p -> Fresh n (Exp (Ann a n) n p)
 anormalAllVars a_fresh xx
  = do   (bs, x)  <- pullSubExps a_fresh xx
-        -- Get the union of all the variables
-        let !allNames = varsOfLets bs x
-        -- and rename the outside binds if they are used
-        (bs',x') <- renames allNames bs [] x
-
-        -- Tag the result with the union of the original bindings (bs)
-        -- as well as the new names of the bindings.
-        let !allNames' = allNames <> Set.fromList (fmap fst bs')
-        let !ret       = makeLets (a_fresh, allNames') bs' x'
+
+        (ret,_) <- foldM insertBinding (x, snd $ annotOfExp x) (reverse bs)
         return ret
 
  where
-  -- All the variables inside the bindings
-  varsOfLets bs x
-   = Set.unions
-   $ fmap (snd.annotOfExp)
-   $ x : fmap snd bs
-
-  -- The sub-expressions of x might already have let bindings, and we're
-  -- going to pull those out.
-  --
-  -- However, this actually changes the scope of those let bindings
-  -- as they will now encompass any bindings in other sub-expressions.
-  --
-  -- If the same name is used in multiple places, we'll introduce shadowing, which is
-  -- outlawed by the type checker.
-  --
-  -- So, we look at each binding, and if it would shadow something, we rename it.
-  renames _ [] seen  x
-   = return (seen, x)
-
-  renames allNames ((n,b):bs) seen x
-   -- Check if n is used anywhere else, including as a later binding name
-   |  n `Set.member` (allNames <> Set.fromList (fmap fst bs))
```

Comment:
I think this was a large part of it - reconstructing this set every iteration, just to check membership. I wonder whether you could write a rewrite rule: `member n (a <> b) = member n a || member n b`